[DOC] Add Kafka Connect Build documentation (#4183)
strimzi/strimzi-kafka-operator: documentation/modules/deploying/proc-deploy-kafka-connect-using-kafka-connect-build.adoc

// Module included in the following assemblies:
//
// assembly-using-kafka-connect-with-plugins.adoc

[id='creating-new-image-using-kafka-connect-build-{context}']
= Creating new container image automatically using Strimzi

> This procedure shows how to have Strimzi automatically build a new container image with additional connectors.

Reviewer: ```suggestion
This procedure shows how to configure Kafka Connect so that Strimzi automatically builds a new container image with additional connectors.
```

> You can define the connector plugins as part of the `KafkaConnect` custom resource in the `.spec.build.plugins`.

Reviewer: ```suggestion
You define the connector plugins using the `.spec.build.plugins` proeperty of the `KafkaConnect` custom resource.
```

> Strimzi will automatically download them, add them into a new container image.

Reviewer: ```suggestion
Strimzi will automatically download and add the connector plugins into a new container image.
```

> The container will be pushed into the container repository specified in `.spec.build.output` and autoamtically used in the Kafka Connect deployment.

Reviewer: ```suggestion
The container is pushed into the container repository specified in `.spec.build.output` and automatically used in the Kafka Connect deployment.
```

.Prerequisites

* xref:deploying-cluster-operator-str[The Cluster Operator must be deployed.]

> * Container registry where the newly built container can be pushed

Reviewer: ```suggestion
* A container registry

You need to provide your own container registry where images can be pushed to, stored, and pulled from.
Strimzi supports private container registries as well as public registries such as link:https://quay.io/[Quay^] or link:https://hub.docker.com//[Docker Hub^].
```

.Procedure

> . Prepare the `KafkaConnect` custom resource with the container registry configured in `.spec.build.output` and additional connectors configured in `.spec.build.plugins`:

Reviewer: ```suggestion
. Configure the `KafkaConnect` custom resource by specifying the container registry in `.spec.build.output`, and additional connectors in `.spec.build.plugins`:
```

> . Prepare the `KafkaConnect` custom resource with the container registry configured in `.spec.build.output` and additional connectors configured in `.spec.build.plugins`:

Reviewer: ```suggestion
. Edit the `KafkaConnect` resource to specify the container registry using the `.spec.build.output` property, and additional connectors using `.spec.build.plugins`:
```

+
[source,yaml,subs=attributes+,options="nowrap"]
----
apiVersion: {KafkaConnectApiVersion}
kind: KafkaConnect
metadata:
  name: my-connect-cluster
spec: # <1>
  #...
  build:
    output: # <2>
      type: docker
      image: my-registry.io/my-org/my-connect-cluster:latest
      pushSecret: my-registry-credentials
    plugins: # <3>
      - name: debezium-postgres-connector
        artifacts:
          - type: tgz
            url: https://repo1.maven.org/maven2/io/debezium/debezium-connector-postgres/1.3.1.Final/debezium-connector-postgres-1.3.1.Final-plugin.tar.gz

>             sha512sum: 962a12151bdf9a5a30627eebac739955a4fd95a08d373b86bdcea2b4d0c27dd6e1edd5cb548045e115e33a9e69b1b2a352bee24df035a0447cb820077af00c03

Reviewer: same comment on shrinking these if possible

      - name: camel-telegram
        artifacts:
          - type: tgz
            url: https://repo.maven.apache.org/maven2/org/apache/camel/kafkaconnector/camel-telegram-kafka-connector/0.7.0/camel-telegram-kafka-connector-0.7.0-package.tar.gz
            sha512sum: a9b1ac63e3284bea7836d7d24d84208c49cdf5600070e6bd1535de654f6920b74ad950d51733e8020bf4187870699819f54ef5859c7846ee4081507f48873479
  #...
----
<1> link:{BookURLUsing}#type-KafkaConnectSpec-reference[The specification for the Kafka Connect cluster^].

> <2> Configuration of the container registry where the new image should be pushed.

Reviewer: ```suggestion
<2> (Required) Configuration of the container registry where the new image are pushed.
```

<3> List of connector plugins which should be added to the container image and their artifacts.

> . Create the `KafkaConnect` custom resource from the YAML file:

Reviewer: ```suggestion
. Create or update the resource:
```

+
[source,subs="+quotes"]
----

> $ kubectl apply -f <YAMLfile>

Reviewer: ```suggestion
$ kubectl apply -f _KAFKA-CONNECT-CONFIG-FILE_
```

----

> . Wait for the new container image to be built and for the Kafka Connect cluster to be deployed.

Reviewer: ```suggestion
. Wait for the new container image to build, and for the Kafka Connect cluster to be deployed.
```

> . Use the Kafka Connect REST API or the KafkaConnector custom resources to use the connector plugins you added.

Reviewer: ```suggestion
. Use the Kafka Connect REST API or the `KafkaConnector` custom resources to use the connector plugins you added.
```

.Additional resources

See the _Using Strimzi_ guide for more information on:

> * link:{BookURLUsing}#type-Build-reference[Kafka Connect Build reference^]

Reviewer: ```suggestion
* link:{BookURLUsing}#type-Build-reference[Kafka Connect `Build` schema reference^]
```
